src/TerraformEnvironmentStage.groovy

```diff
@@ -1,14 +1,14 @@
 class TerraformEnvironmentStage implements Stage {
-    private Jenkinsfile jenkinsfile
-    private String environment
-    private StageDecorations decorations
-    private TerraformInitCommand initCommand
-    private TerraformPlanCommand planCommand
-    private TerraformApplyCommand applyCommand
-    private localPlugins
+    protected Jenkinsfile jenkinsfile
+    protected String environment
+    protected StageDecorations decorations
+    protected TerraformInitCommand initCommand
+    protected TerraformPlanCommand planCommand
+    protected TerraformApplyCommand applyCommand
+    protected localPlugins
 
-    private static final DEFAULT_PLUGINS = [ new ConditionalApplyPlugin(), new ConfirmApplyPlugin(), new DefaultEnvironmentPlugin() ]
-    private static globalPlugins = DEFAULT_PLUGINS.clone()
+    protected static final DEFAULT_PLUGINS = [ new ConditionalApplyPlugin(), new ConfirmApplyPlugin(), new DefaultEnvironmentPlugin() ]
+    protected static globalPlugins = DEFAULT_PLUGINS.clone()
 
```

Comment:
Now that you've done that, you just highlighted the fact that many of these variables never needed to be instance variables to begin with.

Rather than make initCommand/planCommand/applyCommand protected, you could simply delete them, and make them local only to the pipelineConfiguration method.  In which case TerraformDestroy would have no idea about them (which is good) and would simply define it's own.

environment already has a public getEnvironment() method, it would not need to be made public.  jenkinsfile can be accessed globally, it too would not need to be made protected.

applyPlugins manipulates the plugins as needed, none of the plugin variables would need to be made protected.

Reply:
I feel like you've highlighted a code smell as well.  I think you're showing that maybe pipelineConfiguration should perhaps be implemented in some other way.  In fact, you might have just hit the nail on the head.  If pipelineConfiguration were defined by a class/object that could be swapped out, that might in fact solve *everything*.  We wouldn't need a new TerraformDestoryStage at all.